### ISSUE TITLE
Wrong change ids in autoupdates

### ISSUE THREAD

**Issue by FinnStutzenstein** [OpenSlides/openslides3-autoupdate-service#65]
The following sequence in the stream resulted in missing autoupdates:

```
      4086) 1) "1598967102649-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:860\":{\"id\":860,\"username\":\"user827\",\"title\":\"\",\"first_name\":\"user827\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":true,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user827\",\"session_auth_hash\":\"a0b04709c0b050d835bdc042c19f8cc53175a53b\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681074}"
      4087) 1) "1598967102652-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:251\":{\"id\":251,\"username\":\"user218\",\"title\":\"\",\"first_name\":\"user218\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":false,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user218\",\"session_auth_hash\":\"4958e9f8fadf0c8ec54a71df837b16acba627cb2\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681087}"
      4088) 1) "1598967102653-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:697\":{\"id\":697,\"username\":\"user664\",\"title\":\"\",\"first_name\":\"user664\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":true,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user664\",\"session_auth_hash\":\"34e7a7c39bde1839015392ef004d4fdcdf162071\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681088}"
      4089) 1) "1598967102658-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:772\":{\"id\":772,\"username\":\"user739\",\"title\":\"\",\"first_name\":\"user739\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":true,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user739\",\"session_auth_hash\":\"d174ac390c594ba3fe112dac8de10619a5fedaa1\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681091}"
      4090) 1) "1598967102659-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:112\":{\"id\":112,\"username\":\"user79\",\"title\":\"\",\"first_name\":\"user79\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":true,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user79\",\"session_auth_hash\":\"05b102c42a9228147caf539064d6128df34c9528\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681090}"
      4091) 1) "1598967102659-1"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:668\":{\"id\":668,\"username\":\"user635\",\"title\":\"\",\"first_name\":\"user635\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":false,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user635\",\"session_auth_hash\":\"58b590fe21a57ab9af4a56fc7ef8464488cbc343\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681089}"
      4092) 1) "1598967102668-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:554\":{\"id\":554,\"username\":\"user521\",\"title\":\"\",\"first_name\":\"user521\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":false,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user521\",\"session_auth_hash\":\"1feed6d0273d696bc7267625ed607a58a092624f\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681092}"
      4093) 1) "1598967102687-0"
            2) 1) "content"
               2) "{\"elements\":{\"users/user:1007\":{\"id\":1007,\"username\":\"user974\",\"title\":\"\",\"first_name\":\"user974\",\"last_name\":\"\",\"structure_level\":\"\",\"number\":\"\",\"about_me\":\"\",\"groups_id\":[3,4,5],\"is_present\":false,\"is_committee\":false,\"vote_weight\":\"1.000000\",\"gender\":\"\",\"email\":\"\",\"last_email_send\":null,\"comment\":\"\",\"is_active\":true,\"auth_type\":\"default\",\"default_password\":\"user974\",\"session_auth_hash\":\"c891a5092f9d1dc72756f953cdee06dfe48abe99\",\"_no_delete_on_restriction\":false}},\"change_id\":142258681093}"
```

```
received {changed: {…}, deleted: {…}, from_change_id: 142258681086, to_change_id: 142258681088, all_data: false}
received {changed: {…}, deleted: {…}, from_change_id: 142258681088, to_change_id: 142258681089, all_data: false}
received {changed: {…}, deleted: {…}, from_change_id: 142258681091, to_change_id: 142258681092, all_data: false}
received {changed: {…}, deleted: {…}, from_change_id: 142258681092, to_change_id: 142258681093, all_data: false}
```
The autoupdate for 142258681090 is missing. (Note: the from change id is exclusive)

But looking in the autoupdate from 88 to 89, the user 112 (change id 90) is included there:
```
{
  "changed": {
    "users/user": [
      {
        "about_me": "",
        "auth_type": "default",
        "comment": "",
        "default_password": "user739",
        "email": "",
        "first_name": "user739",
        "gender": "",
        "groups_id": [
          3,
          4,
          5
        ],
        "id": 772,
        "is_active": true,
        "is_committee": false,
        "is_present": true,
        "last_email_send": null,
        "last_name": "",
        "number": "",
        "structure_level": "",
        "title": "",
        "username": "user739",
        "vote_weight": 1
      },
      {
        "about_me": "",
        "auth_type": "default",
        "comment": "",
        "default_password": "user79",
        "email": "",
        "first_name": "user79",
        "gender": "",
        "groups_id": [
          3,
          4,
          5
        ],
        "id": 112,
        "is_active": true,
        "is_committee": false,
        "is_present": true,
        "last_email_send": null,
        "last_name": "",
        "number": "",
        "structure_level": "",
        "title": "",
        "username": "user79",
        "vote_weight": 1
      },
      {
        "about_me": "",
        "auth_type": "default",
        "comment": "",
        "default_password": "user635",
        "email": "",
        "first_name": "user635",
        "gender": "",
        "groups_id": [
          3,
          4,
          5
        ],
        "id": 668,
        "is_active": true,
        "is_committee": false,
        "is_present": false,
        "last_email_send": null,
        "last_name": "",
        "number": "",
        "structure_level": "",
        "title": "",
        "username": "user635",
        "vote_weight": 1
      }
    ]
  },
  "deleted": {},
  "from_change_id": 142258681088,
  "to_change_id": 142258681089,
  "all_data": false
}
```
In this autoupdate users with ids 772, 112 and 668 are included which corresponds to change ids 89 to 91. This seems right, but the change ids in the autoupdate are wrong. If this autoupdate had have the to_change_id of 91, everything would be right: it contains 88..91, so the interval would be ]88,91] reported to the client.

**Comment by ostcar** [comment]
This is strange. The "from_change_id" value is actually a fake value. It is just the value of the last autoupdate-message:

https://github.com/OpenSlides/openslides3-autoupdate-service/blob/ff54ba7ff3e0ddc6db65558e22b48c59575d9631/internal/autoupdate/http.go#L48

I can not see from the code, how it is possible, that the "from_change_id" value is different to the "to_change_id" value of the previous autoupdate-message.

Could you tell me, how I can reproduce this? It is not enough to insert the xadd values to redis. If a change_id is missing, the autoupdate-service calls `ZRANGEBYSCORE change_id ...` to receive the missing data.

**Comment by FinnStutzenstein** [comment]
I do not know. This happend under higher load (~100 autoupdates per second). Maybe you can write a script, that just inserts autoupdates into redis and test...

**Comment by ostcar** [comment]
I tested it as good as I can and it works as expected.

Maybe one Message from the autoupdate-service to the client got lost? Is there a log message from the autoupdate-service, that a message could not be send? Is it possible, that for any reason the client not print the `received` console-message?

Without further information, I can not help. Please create a test-setting with a curl request so we can be sure, that the bug is actually in the autoupdate-service.

**Comment by FinnStutzenstein** [comment]
The point is, that actally there is nothing lost. All data from the redis stream is present in the updates, but the change ids are wrong. I can recreate it with a autoupdate rate of 10/sec. No, there are no error messages. I'll send you a possibility to see it later via email.

**Comment by ostcar** [comment]
This should be fixed with #66. Please reopen if not.
